### PR TITLE
fix(protocol-designer): fix labware tools filtering and expand/collapse behavior

### PR DESCRIPTION
**Pull request by ncdiehl11**
# Overview

This PR addresses several functional issues with our LabwareTools component, produced when adding labware to the starting deck state. Here, I add filtering to populated expanded categories, allow independent expand/collapse toggling for multiple categories simultaneously, and auto expand/collapse all categories based on the state of the current search term.

NOTE: This component is pretty bloated as is, and I think we can address condensing filtering logic and improving performance in a followup. Also, I notice a ton of rerenders after mounting— I will address in a followup.

Closes RQA-3590
Closes RQA-3339

https://github.com/user-attachments/assets/496045fa-9a28-44cd-82e8-f4c94cfd03d9

## Test Plan and Hands on Testing

- navigate to starting deck and add hardware/labware to a slot
- click labware tab
- upload a custom definition or two and verify that custom labware category shows and is expanded
- click around categories and verify that they can be expanded/collapsed independently
- begin typing a search term and verify that only populated categories show, and all categories are expanded
- verify that with search term still present, each category can be collapsed/expanded independently
- delete the search term either by backspacing all the way or clicking the 'X' icon, and verify that all categories appear and all collapsed

## Changelog

- fix filtering for populated categories
- add state that handles all categories expanded/collapsed state simultaneously
- pluralize category titles

## Review requests

- see test plan

## Risk assessment

medium

### CODE REVIEW

File: protocol-designer/src/assets/localization/en/starting_deck_state.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "adapter_compatible_lab": "Adapter compatible labware",
-  "adapter": "Adapter",
+  "adapter": "Adapters",
```
Comment (koji) [inline protocol-designer/src/assets/localization/en/starting_deck_state.json:3]: nit

```suggestion
  "adapters": "Adapters",
```

File: protocol-designer/src/assets/localization/en/starting_deck_state.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
   "add_module": "Add a module",
   "add_rest": "Add labware and liquids to complete deck setup",
   "alter_pause": "You may also need to alter the time you pause while your magnet is engaged.",
-  "aluminumBlock": "Aluminum block",
+  "aluminumBlock": "Aluminum blocks",
```
Comment (koji) [inline protocol-designer/src/assets/localization/en/starting_deck_state.json:12]: nit

```suggestion
  "aluminumBlocks": "Aluminum blocks",
```

File: protocol-designer/src/assets/localization/en/starting_deck_state.json [hunk 3]
```diff
@@ -47,16 +47,16 @@
   "protocol_starting_deck": "Protocol starting deck",
   "read_more_gen1_gen2": "Read more about the differences between GEN1 and GEN2 Magnetic Modules",
   "rename_lab": "Rename labware",
-  "reservoir": "Reservoir",
+  "reservoir": "Reservoirs",
```
Comment (koji) [inline protocol-designer/src/assets/localization/en/starting_deck_state.json:50]: nit

```suggestion
  "reservoirs": "Reservoirs",
```

File: protocol-designer/src/assets/localization/en/starting_deck_state.json [hunk 4]
```diff
@@ -47,16 +47,16 @@
...
   "shift_click_to_select_all": "Shift + Click to select all",
   "starting_deck_state": "Starting deck state",
   "tc_slots_occupied_flex": "The Thermocycler needs slots A1 and B1. Slot A1 is occupied",
   "tc_slots_occupied_ot2": "The Thermocycler needs slots 7, 8, 10, and 11. One or more of those slots is occupied",
-  "tipRack": "Tip rack",
+  "tipRack": "Tip racks",
```
Comment (koji) [inline protocol-designer/src/assets/localization/en/starting_deck_state.json:55]: nit 

```suggestion
  "tipRacks": "Tip racks",
```

File: protocol-designer/src/assets/localization/en/starting_deck_state.json [hunk 5]
```diff
@@ -47,16 +47,16 @@
...
   "trash_required": "A trash bin or waste chute is required",
-  "tubeRack": "Tube rack",
+  "tubeRack": "Tube racks",
```
Comment (koji) [inline protocol-designer/src/assets/localization/en/starting_deck_state.json:57]: nit

```suggestion
  "tubeRacks": "Tube racks",
```

File: protocol-designer/src/assets/localization/en/starting_deck_state.json [hunk 6]
```diff
@@ -47,16 +47,16 @@
...
   "untitled_protocol": "Untitled protocol",
   "upload_custom_labware": "Upload custom labware",
   "we_added_hardware": "We've added your deck hardware!",
-  "wellPlate": "Well plate"
+  "wellPlate": "Well plates"
```
Comment (koji) [inline protocol-designer/src/assets/localization/en/starting_deck_state.json:61]: nit 

```suggestion
  "wellPlates": "Well plates"
```

File: protocol-designer/src/pages/Designer/DeckSetup/LabwareTools.tsx [hunk 7]
```diff
@@ -87,11 +87,30 @@ export function LabwareTools(props: LabwareToolsProps): JSX.Element {
...
+  const allCategoriesExpanded = allOrderedCategories.reduce<
+    Record<string, boolean>
+  >((acc, category) => {
+    return { ...acc, [category]: true }
+  }, {})
+  const allCategoriesCollapsed = allOrderedCategories.reduce<
+    Record<string, boolean>
+  >((acc, category) => {
+    return { ...acc, [category]: false }
+  }, {})
```
Comment (koji) [inline protocol-designer/src/pages/Designer/DeckSetup/LabwareTools.tsx]: nit
```ts
const allCategoriesExpanded = allOrderedCategories.reduce<Record<string, boolean>>(
  (acc, category) => ({ ...acc, [category]: true }),
  {}
);

const allCategoriesCollapsed = allOrderedCategories.reduce<Record<string, boolean>>(
  (acc, category) => ({ ...acc, [category]: false }),
  {}
);
```
or 

```ts
const createCategoryState = (categories: string[], state: boolean): Record<string, boolean> => 
  categories.reduce<Record<string, boolean>>((acc, category) => ({ ...acc, [category]: state }), {});

const allCategoriesExpanded = createCategoryState(allOrderedCategories, true);
const allCategoriesCollapsed = createCategoryState(allOrderedCategories, false);
```

File: protocol-designer/src/pages/Designer/DeckSetup/LabwareTools.tsx [hunk 8]
```diff
@@ -87,11 +87,30 @@ export function LabwareTools(props: LabwareToolsProps): JSX.Element {
...
+  const [areCategoriesExpanded, setAreCategoriesExpanded] = React.useState<
+    Record<string, boolean>
+  >(allCategoriesCollapsed)
   const [searchTerm, setSearchTerm] = React.useState<string>('')
 
+  React.useEffect(() => {
```
Comment (koji) [inline protocol-designer/src/pages/Designer/DeckSetup/LabwareTools.tsx]:
```suggestion
import { useEffect } from 'react'  

  useEffect(() => {
```

File: protocol-designer/src/assets/localization/en/starting_deck_state.json (hunk 1 above)
Comment (jerader) [reply protocol-designer/src/assets/localization/en/starting_deck_state.json:3]: i think these are all dependent on the category name which is why they are not plural

File: protocol-designer/src/pages/Designer/DeckSetup/LabwareTools.tsx [hunk 9]
```diff
@@ -186,13 +205,22 @@ export function LabwareTools(props: LabwareToolsProps): JSX.Element {
 
   const populatedCategories: { [category: string]: boolean } = React.useMemo(
```
Comment (koji) [inline protocol-designer/src/pages/Designer/DeckSetup/LabwareTools.tsx]: could you remove `React.` by importing useMemo?

File: protocol-designer/src/pages/Designer/DeckSetup/LabwareTools.tsx [hunk 10]
```diff
@@ -305,144 +343,136 @@ export function LabwareTools(props: LabwareToolsProps): JSX.Element {
...
-                                              />
-                                            )
-                                          }
-                                        )
-                                      : getLabwareCompatibleWithAdapter(
-                                          loadName
-                                        ).map(nestedDefUri => {
-                                          const nestedDef = defs[nestedDefUri]
+                      return searchFilter(labwareDef.metadata.displayName) ? (
+                        <React.Fragment
```
Comment (koji) [inline protocol-designer/src/pages/Designer/DeckSetup/LabwareTools.tsx]:
```suggestion
import { Fragment } from 'react'

                        <Fragment
```

**Review by koji — APPROVED** [review]
sandbox worked as expected

**Comment by jerader** [comment]
| Also, I notice a ton of rerenders after mounting— I will address in a followup

@ncdiehl11 by rerendering, do you mean deck setup tools rerenders a bunch of times when you zoom into a slot? if so, that's a known issue: https://opentrons.atlassian.net/browse/AUTH-769

**Review by jerader — APPROVED** [review]
makes sense to me. might be worth in a followup when we refactor/clean up these components to add setSeachTerm/searchTerm and areCategoriesExpanded/setAreCategoriesExpanded to redux. Thanks for fixing this!

**Comment by ncdiehl11** [comment]
> | Also, I notice a ton of rerenders after mounting— I will address in a followup
> 
> @ncdiehl11 by rerendering, do you mean deck setup tools rerenders a bunch of times when you zoom into a slot? if so, that's a known issue: https://opentrons.atlassian.net/browse/AUTH-769

yes. I think it may be introduced in this component. We can get to it soon

**Pull request merged**